crucible-llvm/src/Lang/Crucible/LLVM.hs

```diff
@@ -25,43 +31,157 @@ module Lang.Crucible.LLVM
+-- | Lazily register the named function that is defnied in the given module
+--   translation. This will delay actually translating the function until it
+--   is called. This done by first installing a bootstrapping override that
+--   will peform the actual translation when first invoked, and then will backpatch
+--   its own references to point to the translated function.
+registerLazyModuleFn ::
+   (1 <= ArchWidth arch, HasPtrWidth (ArchWidth arch), IsSymInterface sym) =>
+   (LLVMTranslationWarning -> IO ()) ->
+   LLVMContext arch ->
+   ModuleTranslation arch ->
+   L.Symbol ->
+   OverrideSim p sym LLVM rtp l a ()
+registerLazyModuleFn handleWarning llvm_ctx mtrans sym =
+  case filter (\def -> L.defName def == sym) (L.modDefines (mtrans ^. modTransModule)) of
+    [def] ->
+      let ?lc = llvm_ctx^.llvmTypeCtx in
+      let decl = declareFromDefine def
+          (L.Symbol fstr) = L.decName decl
+          halloc = mtrans ^. modTransHalloc
+       in llvmDeclToFunHandleRepr' decl $ \argTys retTy ->
+           do let fn_name = functionNameFromText $ Text.pack (fstr ++ "_bootstrap")
+              h <- liftIO $ mkHandle' halloc fn_name argTys retTy
+
+              -- Bind the function handle we just created to the following bootstrapping code,
+              -- which actually translates the function on its first execution and patches up
+              -- behind itself.
+              bindFnHandle h
+                $ UseOverride
+                $ mkOverride' fn_name retTy
+                $ -- This inner action defines what to do when this function is called for the
+                  -- first time.  We actually translate the function and install it, which
+                  -- overrites the global symbol.
+                  do SomeHandle h' <- registerModuleFn handleWarning llvm_ctx mtrans sym
+                     case (testEquality argTys (handleArgTypes h'),
+                           testEquality retTy (handleReturnType h')) of
+                        (Nothing, _) -> panic ("Argument type mismatch when bootstraping function: " ++ fstr)
+                                               [ show argTys, show h ]
+                        (_, Nothing) -> panic ("Return type mismatch when bootstrapping function: " ++ fstr)
+                                               [ show retTy, show h ]
+                        (Just Refl, Just Refl) ->
+                           do -- Rebind the old handle to call directly into the new handle.
+                              -- This makes sure that even if the handle related to the global symbol
+                              -- has previously been resolved, the actual code for this function
+                              -- is correctly executed.
+                              bindFnHandle h
+                                $ UseOverride
+                                $ mkOverride' fn_name retTy
+                                $ (regValue <$> (callFnVal (HandleFnVal h') =<< getOverrideArgs))
```

Comment:
I'm trying to figure out what implications this will have on `crux-llvm` profiling reports. If I understand how this works correctly, a profiling report will now display twice as many functions as it used to, since invoking a C function is tantamount to calling two functions on the Crucible side: one named `foo_bootstrap` and one named `foo`. Is that right? That might add some unfortunate clutter.

One alternative to the double-override approach would be to:

1. Give the bootstrapping function that name "`foo`" rather than "`foo_bootstrap`"
2. Inline the parts of `registerModuleFn` that construct a `CFG` for the translated function
3. Re-bind the bootstrapping function handle to a `UseCFG` rather than a `UseOverride`

That way, the profiling reports wouldn't have to mention `_bootstrap` names at all. The first invocation of the function would likely take up more time on the report due to the need to translate the CFG, but perhaps this is an acceptable compromise. Can you think of any other downsides to this?

Reply:
Yeah, the profiling data will be a bit messier. I think that's probably OK? I guess we could make it a configurable to use lazy or eager loading if users want a more straightforward profile.

I'm not sure the alternative you suggest would be better, but maybe I'm missing some advantage you have in mind?

Reply:
> I'm not sure the alternative you suggest would be better, but maybe I'm missing some advantage you have in mind?

The difference in my approach is that when we invoke the bootstrap function for the first time, instead of creating a new function handle for the translated function, we just simply re-register the bootstrap function's handle with the CFG of the translated function. That way, there's no need to have two handles for each LLVM function. Instead, we would have a single handle per function whose `FnState` will be registered to different things depending on whether you have invoked it once or not.

Reply:
Well, I guess you'd still have two handles for each function, since the CFG has its own handle. But you'd only need to _register_ the handle for the bootstrapping function in the simulator, so from the perspective of the profiler, there is only one thing that shows up in the profiler for each function.

Reply:
I guess we could change the API for translation so that it could take a function handle instead of inventing its own.  That would mean we only have to rebind a single function handle.

I'm still not sure if it is better to hide the lazy loading from profiling or not.  I think having the time spent on loading and translating the code just be silently included in the profile might be pretty confusing, actually.

Reply:
> I think having the time spent on loading and translating the code just be silently included in the profile might be pretty confusing, actually.

True. I'm not sure which of the two approach's downsides would be worse.

A third possible approach: instead of registering boostrap functions for every LLVM function at the start of simulation, don't register anything at all. Later, when running an `LLVM_LoadHandle` in the memory model, there is a check [here](https://github.com/GaloisInc/crucible/blob/b558ba2341b11c88e668ed3f772f4d1960275229/crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs#L440-L468) to see if the loaded handle has been registered in the simulator. Currently, this check will throw an assertion failure if that is not the case, but we could instead perform the lazy translation and registration there. This would ensure both that each function is only registered once and that the time spent during profiling only measures execution time, not translation time. (This is similar to the approach to profiling that we use in a `macaw`-based verifier that also supports lazy translation.)

I haven't fully thought through what code changes would need to happen to make this work. For instance, you might need to plumb through some additional state to the memory model to support lazy translation. But I think this would address both issues, provided it's feasible. If it isn't, I certainly don't want this to hold up the PR.